aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/customize/glacier/GlacierDecorator.kt

```diff
@@ -6,21 +6,206 @@
+            it + listOf(GlacierApiVersionCustomization(codegenContext))
+        }
+
+    override fun operationRuntimePluginCustomizations(
+        codegenContext: ClientCodegenContext,
+        operation: OperationShape,
+        baseCustomizations: List<OperationRuntimePluginCustomization>,
+    ): List<OperationRuntimePluginCustomization> =
+        baseCustomizations.letIf(codegenContext.smithyRuntimeMode.generateOrchestrator) {
+            it + listOf(GlacierOperationInterceptorsCustomization(codegenContext))
+        }
+
+    override fun requestSerializerCustomizations(
+        codegenContext: ClientCodegenContext,
+        operation: OperationShape,
+        baseCustomizations: List<RequestSerializerCustomization>,
+    ): List<RequestSerializerCustomization> = baseCustomizations.letIf(operation.requiresTreeHashHeader()) {
+        it + listOf(GlacierTreeHashHeaderCustomization(codegenContext))
+    }
+}
+
+/** Implements the `GlacierAccountId` trait for inputs that have an `account_id` field */
+private class GlacierAccountIdCustomization(private val codegenContext: ClientCodegenContext) :
+    StructureCustomization() {
+    override fun section(section: StructureSection): Writable = writable {
+        if (section is StructureSection.AdditionalTraitImpls && section.shape.inputWithAccountId()) {
+            val inlineModule = inlineModule(codegenContext.runtimeConfig)
+            rustTemplate(
+                """
+                impl #{GlacierAccountId} for ${section.structName} {
+                    fn account_id_mut(&mut self) -> &mut Option<String> {
+                        &mut self.account_id
+                    }
+                }
+                """,
+                "GlacierAccountId" to inlineModule.resolve("GlacierAccountId"),
+            )
+        }
+    }
 }
+
+/** Adds the `x-amz-glacier-version` header to all requests */
+private class GlacierApiVersionCustomization(private val codegenContext: ClientCodegenContext) :
+    ServiceRuntimePluginCustomization() {
+    override fun section(section: ServiceRuntimePluginSection): Writable = writable {
+        if (section is ServiceRuntimePluginSection.AdditionalConfig) {
+            val apiVersion = codegenContext.serviceShape.version
+            section.registerInterceptor(codegenContext.runtimeConfig, this) {
+                rustTemplate(
+                    "#{Interceptor}::new(${apiVersion.dq()})",
+                    "Interceptor" to inlineModule(codegenContext.runtimeConfig).resolve("GlacierApiVersionInterceptor"),
+                )
+            }
+        }
+    }
+}
+
+/**
+ * Adds two interceptors:
+ * 1. `GlacierAccountIdAutofillInterceptor`: Uses the `GlacierAccountId` trait to correctly autofill the account ID field
+ * 2. `GlacierSetPrecomputedSignableBodyInterceptor`: Reuses the tree hash computation during signing rather than allowing
+ *    the `aws-sigv4` module to recalculate the payload hash.
```

Comment:
Another chance for the orc to be faster than the old client :D

Reply:
This is all adapted from the middleware implementation. It already had this.